Speed up calculate Qparams for per-channel observers

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30485 Speed up calculate Qparams for per-channel observers**

Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

Commit:
Speed up calculate Qparams for per-channel observers

Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -195,27 +165,26 @@ def _calculate_qparams(self, min_val, max_val):
+        if self.qscheme == torch.per_tensor_symmetric or self.qscheme == torch.per_channel_symmetric:
+            max_val = torch.max(-min_val, max_val)
+            scale = max_val / (float(qmax - qmin) / 2)
+            scale = torch.max(scale, torch.tensor(1.0).new_full(scale.size(), self.eps, dtype=torch.float32))
```
Why is this comparison needed? We do not need scale to be greater than 1.0.

hx89 replied:
This is to make sure scale is greater than self.eps. Since scale is a tensor we need to create a tensor with value self.eps and the same shape as scale to compare with scale using torch.max(). I used new_full() on a dummy tensor with value 1.0 here to create such tensor, but there might be other ways to do it.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -195,27 +165,26 @@ def _calculate_qparams(self, min_val, max_val):
-                zero_point = min(qmax, zero_point)
-                zero_point = int(zero_point)
+            scale = (max_val - min_val) / float(qmax - qmin)
+            scale = torch.max(scale, torch.tensor(1.0).new_full(scale.size(), self.eps, dtype=torch.float32))
```
Similar questions, can you explain why this is needed?

hx89 replied:
It's the same reason as above.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -195,27 +165,26 @@ def _calculate_qparams(self, min_val, max_val):
+            scale = (max_val - min_val) / float(qmax - qmin)
+            scale = torch.max(scale, torch.tensor(1.0).new_full(scale.size(), self.eps, dtype=torch.float32))
+            zero_point = qmin - torch.round(min_val / scale)
+            zero_point = torch.max(zero_point, torch.tensor(1.0).new_full(zero_point.size(), qmin, dtype=torch.float32))
```
Again, do we need the torch.min and torch.max here?

hx89 replied:
It's the same reason as above.

raghuramank100 reviewed (changes requested):
Looks great, a few questions:

hx89 commented:
Thanks @raghuramank100 for reviewing! I've addressed your questions.

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -195,27 +165,27 @@ def _calculate_qparams(self, min_val, max_val):
+            scale = (max_val - min_val) / float(qmax - qmin)
+            scale = torch.max(scale, torch.tensor(1.0, device=device).new_full(scale.size(), self.eps, dtype=scale.dtype))
+            zero_point = qmin - torch.round(min_val / scale)
+            zero_point = torch.max(zero_point, torch.tensor(1.0, device=device).new_full(zero_point.size(), qmin, dtype=zero_point.dtype))
```
Zero-point should be in the range of qmin and qmax.  Cleaner way to do this would be (broadcasting is supported)
```python
zero_point  = torch.max(zero_point, torch.tensor(qmin, dtype=zero_point.dtype))
```

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -195,27 +165,27 @@ def _calculate_qparams(self, min_val, max_val):
-                zero_point = min(qmax, zero_point)
-                zero_point = int(zero_point)
+            scale = (max_val - min_val) / float(qmax - qmin)
+            scale = torch.max(scale, torch.tensor(1.0, device=device).new_full(scale.size(), self.eps, dtype=scale.dtype))
```
Similar comments on what I have suggested for zero-point

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit c5e9f59:
* **1/2** broken upstream at merge base b0e7db5 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=b0e7db5b31a093e08e01c1f60eef671528105937&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * Since your merge base is older than `viable/strict`, run these commands:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **1/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c5e9f59c4ff58ed851f8476c86e3376e69fde70d).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3935909) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=73665569))
```
CONFLICT (add/add): Merge conflict in .circleci/validate-docker-version.py 
Auto-merging .circleci/validate-docker-version.py 
CONFLICT (add/add): Merge conflict in .circleci/scripts/should_run_job.sh 
Auto-merging .circleci/scripts/should_run_job.sh 
CONFLICT (add/add): Merge conflict in .circleci/generate_config_yml.py 
Auto-merging .circleci/generate_config_yml.py 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_travis_python.sh 
Auto-merging .circleci/docker/common/install_travis_python.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
Automatic merge failed; fix conflicts and then commit the result. 
```
### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_xla_linux_xenial_py3_6_clang7_build](https://dr.pytorch.org/build-details.html?build_id=73668427)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

raghuramank100 reviewed (approved):
Thanks for the PR!

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]

Commit:
Update on "Speed up calculate Qparams for per-channel observers"


Use vectorization to speed up calculate Qparams for per-channel observers. New implementation is about 1000 times faster.

Test plan:
```
import torch
import time
import numpy as np
from torch.quantization.observer import PerChannelMinMaxObserver

obs = PerChannelMinMaxObserver()
acc_time = 0
X = torch.randn(1000, 10)
obs(X)
for i in range(100):
    start = time.time()
    obs.calculate_qparams()
    acc_time = acc_time + time.time()-start
print(acc_time)
```
Before change:
20.3
After change:
0.017

Task:
https://github.com/pytorch/pytorch/issues/30348#event-2824868602

Differential Revision: [D18711905](https://our.internmc.facebook.com/intern/diff/D18711905/)

[ghstack-poisoned]
